src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -811,12 +811,13 @@ private GetRecordsResult getRecords(CachingShardIterator iterator) {
                     return loadedRecordsResult;
                 }
                 // otherwise compute next iterator (update cache for lazy iterators)
-                CachingShardIterator nextIterator = iterator.getDynamoDbIterator()
-                    .map(iterator::withDynamoDbIterator)
-                    .orElseGet(() -> {
-                        iteratorCache.put(iterator, loadedNextIterator);
-                        return iterator;
-                    });
+                CachingShardIterator nextIterator;
+                if (iterator.getDynamoDbIterator().isPresent()) {
+                    nextIterator = iterator.withDynamoDbIterator(loadedNextIterator);
+                } else {
+                    iteratorCache.put(iterator, loadedNextIterator);
+                    nextIterator = iterator;
+                }
```

Comment:
The previous version of the code was not returning the new loaded iterator which caused streams to get stuck at `TRIM_HORIZON`.